polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2368,6 +2366,76 @@ ValueCategory MLIRScanner::EmitScalarConversion(ValueCategory Src,
   return Res;
 }
 
+ValueCategory MLIRScanner::EmitFloatToBoolConversion(ValueCategory Src) {
+  assert(Src.val.getType().isa<FloatType>() && "Expecting a float value");
+  mlir::OpBuilder SubBuilder(builder.getContext());
+  SubBuilder.setInsertionPointToStart(entryBlock);
+  auto FloatTy = cast<FloatType>(Src.val.getType());
+  auto Zero = SubBuilder.create<ConstantFloatOp>(
+      builder.getUnknownLoc(),
+      mlir::APFloat::getZero(FloatTy.getFloatSemantics()), FloatTy);
+  return Src.FCmpUNE(builder, {Zero, false});
+}
+
+ValueCategory MLIRScanner::EmitPointerToBoolConversion(ValueCategory Src) {
+  if (auto MemRefTy = Src.val.getType().dyn_cast<MemRefType>()) {
+    auto ElementTy = MemRefTy.getElementType();
+    auto AddressSpace = MemRefTy.getMemorySpaceAsInt();
+    Src = {
+        builder.create<polygeist::Memref2PointerOp>(
+            loc, LLVM::LLVMPointerType::get(ElementTy, AddressSpace), Src.val),
+        true};
+  }
+  assert(Src.val.getType().isa<LLVM::LLVMPointerType>() &&
+         "Expecting a pointer");
+  mlir::OpBuilder SubBuilder(builder.getContext());
+  SubBuilder.setInsertionPointToStart(entryBlock);
+  auto Zero = SubBuilder.create<LLVM::NullOp>(builder.getUnknownLoc(),
+                                              Src.val.getType());
+  return {builder.createOrFold<LLVM::ICmpOp>(loc, LLVM::ICmpPredicate::ne,
+                                             Src.val, Zero),
+          false};
+}
+
+ValueCategory MLIRScanner::EmitIntToBoolConversion(ValueCategory Src) {
+  assert(Src.val.getType().isa<IntegerType>() && "Expecting an integer value");
+  // Because of the type rules of C, we often end up computing a
+  // logical value, then zero extending it to int, then wanting it
+  // as a logical value again.  Optimize this common case.
+  if (auto ZI = Src.val.getDefiningOp<arith::ExtUIOp>()) {
```

Comment:
Should we do these kinds of optimizations in an optimization pass after codegen?

Reply:
I think you're right. I'll drop this.